File: LeetCode/18四数之和.go
func fourSum(nums []int, target int) [][]int {
	var ans [][]int
	sort.Ints(nums)
	for i := 0; i < len(nums); i++ {
		if nums[i] > 0 && nums[i] > target {
			break
		}
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		for j := i + 1; j < len(nums); j++ {
			if nums[i]+nums[j] > 0 && nums[i]+nums[j] > target {
				break
			}
			if j > i+1 && nums[j-1] == nums[j] {
				continue
			}
			left, right := j+1, len(nums)-1
			for left < right {
				temp := nums[i] + nums[j] + nums[left] + nums[right]
				if temp < target {
					left++
				} else if temp > target {
					right--
				} else {
					ans = append(ans, []int{nums[i], nums[j], nums[left], nums[right]})
					for left < right && nums[left] == nums[left+1] {
						left++
					}
					for left < right && nums[right-1] == nums[right] {
						right--
					}
					left++
					right--
				}
			}
		}
	}
	return ans
}

File: LeetCode/18四数之和.java
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        Arrays.sort(nums);
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] > 0 && nums[i] > target) break;
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            for (int j = i + 1; j < nums.length; j++) {
                if (nums[i] + nums[j] > 0 && nums[i] + nums[j] > target) break;
                if (j > i + 1 && nums[j] == nums[j - 1]) continue;
                int left = j + 1;
                int right = nums.length - 1;
                while (left < right) {
                    long temp = nums[i] + nums[j] + nums[left] + nums[right];
                    if (temp < target) left++;
                    else if (temp > target) right--;
                    else {
                        ans.add(Arrays.asList(nums[i], nums[j], nums[left], nums[right]));
                        while (left < right && nums[left] == nums[left + 1]) left++;
                        while (left < right && nums[right] == nums[right - 1]) right--;
                        left++;
                        right--;
                    }
                }
            }
        }
        return ans;
    }
}